ksql-parser/src/main/java/io/confluent/ksql/parser/rewrite/CustomTypeRewriter.java

```diff
@@ -0,0 +1,126 @@
+      final Statement statement
+  ) {
+    this.typeRegistry = Objects.requireNonNull(typeRegistry, "typeRegistry");
+    this.statement = Objects.requireNonNull(statement, "statement");
+  }
+
+  public static boolean requiresRewrite(final Statement statement) {
+    return statement instanceof CreateAsSelect
+        || statement instanceof CreateSource;
+  }
+
+  public Statement rewrite() {
+    return (Statement) new StatementRewriter<>(
+        (e, c) -> ExpressionTreeRewriter.rewriteWith(new Plugin(typeRegistry)::process, e)
+    ).rewrite(statement, null);
+  }
+
+  private static final class Plugin
+      extends VisitParentExpressionVisitor<Optional<Expression>, Context<Void>> {
+
+    private final TypeRegistry typeRegistry;
+
+    Plugin(final TypeRegistry typeRegistry) {
+      super(Optional.empty());
+      this.typeRegistry = Objects.requireNonNull(typeRegistry, "typeRegistry");
+    }
+
+    @Override
+    public Optional<Expression> visitType(final Type node, final Context<Void> context) {
+      switch (node.getSqlType().baseType()) {
+        case ARRAY:
+          return visitArray((SqlArray) node.getSqlType(), context);
+        case MAP:
+          return visitMap((SqlMap) node.getSqlType(), context);
+        case STRUCT:
+          return visitStruct((SqlStruct) node.getSqlType(), context);
+        case ALIAS:
+          return visitAlias((SqlTypeAlias) node.getSqlType(), context);
+        default:
+          return Optional.of(node);
+      }
+    }
+
+    private Optional<Expression> visitArray(final SqlArray array, final Context<Void> context) {
+      final SqlType itemType = array.getItemType();
+      final Type inner =
+          (Type) visitType(new Type(itemType), context)
+              .orElseThrow(() -> new KsqlException("Expected type visit to return non-empty!"));
+
+      return Optional.of(new Type(SqlArray.of(inner.getSqlType())));
+    }
+
+    private Optional<Expression> visitMap(final SqlMap map, final Context<Void> context) {
+      final SqlType itemType = map.getValueType();
+      final Type value =
+          (Type) visitType(new Type(itemType), context)
+              .orElseThrow(() -> new KsqlException("Expected type visit to return non-empty!"));
+
+      return Optional.of(new Type(SqlMap.of(value.getSqlType())));
+    }
+
+    private Optional<Expression> visitStruct(final SqlStruct struct, final Context<Void> context) {
+      final SqlStruct.Builder structBuilder = SqlStruct.builder();
+      for (final Field field : struct.getFields()) {
+        final Type fieldType = (Type) visitType(new Type(field.type()), context)
+            .orElseThrow(() -> new KsqlException("Expected type visit to return non-empty!"));
+        structBuilder.field(field.name(), fieldType.getSqlType());
+      }
+      return Optional.of(new Type(structBuilder.build()));
+    }
+
+    private Optional<Expression> visitAlias(final SqlTypeAlias alias, final Context<Void> context) {
```

Comment:
Can we define an alias type that refers to other aliases? If yes, should we check for circular dependencies? :)